feat(service-worker): add option to config when to register sw

JiaLiPassion opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #20970
Some times, angular will start up some long `macroTask` such as a long delay `setTimeout` (in #20970, `firebase auth` will start a `200s` timeout to do token refresh). so `ngZone` will not become stable and will not trigger `service-worker` to do registration.

## What is the new behavior?
I think there are so many cases that when startup, there will be `macroTask` not complete or cost a lot of time, so in this PR, my idea is add a `callback` to let user to be able to choose when to `register`. 

I added a property in `RegistrationOptions`, the sample is here.
https://github.com/DevIntent/angular-example/pull/1

```javascript
@NgModule({
...
imports: [
ServiceWorkerModule.register('/ngsw-worker.js', {enabled: true,
       toInitFactory: AppComponent.getServiceWorkerInitFactory})]
...
```

and `getServiceWorkerInitFactory` will return a promise, user can resolve it by their choice.
the sample code is here, 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
if the idea is ok, I will update the document.

This PR incorporates #26002.

JiaLiPassion committed:
feat(service-worker): allow configuring when the SW is registered

Fixes #20970

Splaktar commented:
Looks like the build failed with:
`Error: Public API differs from golden file. Please run `gulp public-api:update`.`

Splaktar commented:
Would going down a path similar to how preloading lazy loaded routes works be better? 

I.e. define a swRegistrationStrategy parameter that then has a couple of Angular provided defaults (like `registerImmediately` and `registerWhenStable`). Then allow it to also take user defined strategies similar to how [SelectivePreloadingStrategy](https://github.com/angular/angular/blob/master/aio/content/examples/router/src/app/selective-preloading-strategy.ts) implements a custom strategy based on PreloadingStrategy.

JiaLiPassion commented:
@Splaktar , thank you for review, I will update the PR and let you review again.

JiaLiPassion commented:
@Splaktar , I have updated the PR, change the option to 

- registrationStrategy: optional parameter, specify the strategy that when to register service worker, the available options are: 
  - registerWhenStable: this is the default behavior, service worker will register when Application is stable (no microTasks or macroTasks remain).
  - registerImmediately: register immediately without waiting application to become stable.
  - registerDelay:timeout : register after timeout, timeout is a number to specify the delay timeout, for example `registerDelay:5000` means register after 5 seconds.
  - a factory to get Observable : you can also specify a factory which return an Observable, service worker will register when the Observable emit for the first time. 

I also update the sample at https://github.com/DevIntent/angular-example/pull/1, please review.

Splaktar commented on packages/service-worker/src/module.ts:
```diff
@@ -49,7 +50,30 @@ export function ngswAppInitializer(
+        navigator.serviceWorker.register(script, {scope: options.scope});
+      } else if (registrationStrategy.indexOf('registerDelay') !== -1) {
+        const split = registrationStrategy.split(':');
+        const delay = split.length > 1 ? split[1] : 0;
```
Should the default delay be non-zero? Perhaps 5 or 10s? It would be better to give an intelligent default rather than falling back to just be the same as `registerImmediately`.

JiaLiPassion replied:
the `zero` fallback is a little different with `registerImmediately`.
- `registerImmediately` means no delay at all.
- `zero` means `setTimeout(register, 0)`, it will still wait all `microTasks` to finish.  

So I think maybe `zero` is acceptable. What do you think?

Splaktar replied:
Interesting, can you please specify those details about this default delay in the docs?

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
 
 <code-example path="service-worker-getting-started/src/app/app.module.ts" linenums="false" title="src/app/app.module.ts" region="sw-module"> </code-example>
 
+You can pass some options to `register()` method.
```
You can pass some options to the `register()` method.

JiaLiPassion replied:
Got it, I will remove the ` mark.

Splaktar replied:
Sorry, the ` is fine, I just added a "the".

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+      return someService.initData(); // return an Observable, when data is available, we can register service worker.
+    )};
+  </code-pane>
+</code-tabs>
```
This formatting doesn't seem to be rendering properly in GitHub Markdown.

JiaLiPassion replied:
The `<code-tabs>` and `<code-pane>` will be parsed by `dgeni`, I have tested locally, this part can be rendered correctly.

Splaktar replied:
I know that dgeni will code parse comments for addition to docs, but it also parses markdown?

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+You can pass some options to `register()` method.
+- enabled: optional parameter, by default is true, if enabled is false, the module will behave like the browser not support service worker, and service worker will not be registered.
+- scope: optional parameter, to specify the subset of your content that you want the service worker to control.
+- registrationStrategy: optional parameter, specify the strategy that when to register service worker, the available options are: 
```
specify a strategy that determines when to register the service worker

JiaLiPassion replied:
got it.

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+- enabled: optional parameter, by default is true, if enabled is false, the module will behave like the browser not support service worker, and service worker will not be registered.
+- scope: optional parameter, to specify the subset of your content that you want the service worker to control.
+- registrationStrategy: optional parameter, specify the strategy that when to register service worker, the available options are: 
+  - registerWhenStable: this is the default behavior, service worker will register when Application is stable (no microTasks or macroTasks remain).
```
the service worker will register when the application is stable (no microTasks or macroTasks remain).

JiaLiPassion replied:
got it, change `Application` to `application`.

Splaktar replied:
Added two "the"s in there too

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+- scope: optional parameter, to specify the subset of your content that you want the service worker to control.
+- registrationStrategy: optional parameter, specify the strategy that when to register service worker, the available options are: 
+  - registerWhenStable: this is the default behavior, service worker will register when Application is stable (no microTasks or macroTasks remain).
+  - registerImmediately: register immediately without waiting application to become stable.
```
without waiting for the application to become stable

JiaLiPassion replied:
got it, add `the` before application.

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+- registrationStrategy: optional parameter, specify the strategy that when to register service worker, the available options are: 
+  - registerWhenStable: this is the default behavior, service worker will register when Application is stable (no microTasks or macroTasks remain).
+  - registerImmediately: register immediately without waiting application to become stable.
+  - registerDelay:timeout : register after timeout, timeout is a number to specify the delay timeout, for example `registerDelay:5000` means register after 5 seconds.
```
`registerDelay:timeout`: register after the timeout period. `timeout` is the number of milliseconds to delay registration. For example: `registerDelay:5000` would register the service worker after 5 seconds.

JiaLiPassion replied:
got it.

Splaktar commented on aio/content/guide/service-worker-getting-started.md:
```diff
@@ -66,6 +66,31 @@ Add `ServiceWorkerModule` to the `@NgModule` `imports` array. Use the `register(
+  - registerWhenStable: this is the default behavior, service worker will register when Application is stable (no microTasks or macroTasks remain).
+  - registerImmediately: register immediately without waiting application to become stable.
+  - registerDelay:timeout : register after timeout, timeout is a number to specify the delay timeout, for example `registerDelay:5000` means register after 5 seconds.
+  - a factory to get Observable : you can also specify a factory which return an Observable, service worker will register when the Observable emit for the first time. 
```
A factory to get an Observable: You can also specify a factory which returns an Observable. The service worker will be registered the first time that a value is emitted by the Observable.

JiaLiPassion replied:
got it.

Splaktar reviewed (commented):
I like this direction that this is heading 😄

Splaktar commented:
@alxhub we would love some feedback on this approach before more effort is spent on this.

JiaLiPassion commented:
@Splaktar , thank you for the detailed review, I have updated the PR, and about the `markdown format`, I just took a screenshot, please check it, thank you.
<img width="1002" alt="screeshot" src="https://user-images.githubusercontent.com/1442575/35566170-b4226c56-0603-11e8-8e20-068e9270292a.png">

Splaktar commented:
@JiaLiPassion can you please rebase this?

Splaktar commented:
It looks like the last commit broke the build due to TypeScript 3.1.x changes.

JiaLiPassion commented:
Got it, I will fix it today.

JiaLiPassion commented:
@Splaktar, I have updated the source, please review , thanks.

Splaktar commented:
LGTM! CI Tests now pass. Thank you.

gkalpak committed:
refactor(service-worker): DRY up SW registration logic

gkalpak committed:
test(service-worker): add tests for `RegistrationOptions#registrationStrategy`

gkalpak reviewed (commented):
I pushed a couple of commits (addressing my comments, adding tests, etc.). Also rebased this on #26002, because it allows writing public API docs for `RegistrationOptions`.

WDYT, @JiaLiPassion?

JiaLiPassion commented on packages/service-worker/src/module.ts:
```diff
@@ -44,7 +46,33 @@ export function ngswAppInitializer(
+      const observable = options.registrationStrategy();
+      const subscription = observable.subscribe(() => {
+        navigator.serviceWorker.register(script, {scope: options.scope});
+        subscription.unsubscribe();
```
Sure, thanks!

JiaLiPassion commented on packages/service-worker/src/module.ts:
```diff
@@ -44,7 +46,33 @@ export function ngswAppInitializer(
+        whenStable.then(() => navigator.serviceWorker.register(script, {scope: options.scope}));
+      } else if (registrationStrategy === 'registerImmediately') {
+        navigator.serviceWorker.register(script, {scope: options.scope});
+      } else if (registrationStrategy.indexOf('registerDelay') !== -1) {
```
Yes, you are right! Thanks!

JiaLiPassion commented on packages/service-worker/src/module.ts:
```diff
@@ -44,7 +46,33 @@ export function ngswAppInitializer(
+          options.registrationStrategy :
+          'registerWhenStable';
+      if (registrationStrategy === 'registerWhenStable') {
+        whenStable.then(() => navigator.serviceWorker.register(script, {scope: options.scope}));
```
Yes, you are right! Thanks!

JiaLiPassion commented on aio/content/guide/service-worker-config.md:
````diff
@@ -206,3 +206,15 @@ If the field is omitted, it defaults to:
 ]
 ```
+
+## `register options`
````
Got it, Thanks!

JiaLiPassion commented:
@gkalpak, thanks for the review and commits, Your modification LGTM!

gkalpak commented on packages/service-worker/src/module.ts:
```diff
@@ -17,6 +18,7 @@ import {SwUpdate} from './update';
 export abstract class RegistrationOptions {
   scope?: string;
   enabled?: boolean;
+  registrationStrategy?: (() => Observable<any>)|string;
```
FWIW, I have manually verified this works fine with AoT.

gkalpak commented:
Blocked on #26002 (which this PR sits on top of).

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -1,7 +1,14 @@
+export declare abstract class RegistrationOptions {
```
this API doesn't look quite right to me because:

1. we now have two ways of enabling the sw (one via provider and one via `.register` method - and these two APIs don't have the same signature so it does look like two separate things. Why not type `opts` below as `RegistrationOptions`? 
2. as a DI token all you'll see in the bootstrap/ngmodule code is `RegistrationOptions` and it won't be obvious that this is SW options, if anything we should prefix this with `Sw` as we've done with other apis under @angular/service-worker

IgorMinar reviewed (changes requested):
excellent docs! can you please respond to my API concerns? maybe I'm missing something.. or maybe the API does need to be tweaked. once this is resolved, this PR should be good to go.

IgorMinar commented:
and we'll also need a rebase :-(

gkalpak commented:
This is rebased (and blocked) on #26002 (see https://github.com/angular/angular/pull/21842#issuecomment-437435358).
The API concerns are actually related to #26002, so let's first get #26002 merged and then rebase this on master (and get it merged).

IgorMinar commented on tools/public_api_guard/service-worker/service-worker.d.ts:
```diff
@@ -1,5 +1,7 @@
+    registrationStrategy?: (() => Observable<any>) | string;
```
is this a manual edit? seems wrong... can you please take a look?

gkalpak replied:
Hm...it has been fixed in a subsequent commit (faefc1122), but I will backport the change to this one.

IgorMinar reviewed (approved):
thanks! ❤️

this PR is not properly rebased on top of #26002 can you please fix that, close #26002 and send this one for merge instead? that way there is just a single PR to merge.

Commit:
feat(service-worker): expose `SwRegistrationOptions` token to allow runtime config

Previously, the ServiceWorker registration options should be defined as
an object literal (in order for them to be compatible with Ahead-of-Time
compilation), thus making it impossible to base the ServiceWorker
behavior on runtime conditions.
This commit allows specifying the registration options using a regular
provider, which means that it can take advantage of the `useFactory`
option to determine the config at runtime, while still remaining
compatible with AoT compilation.

gkalpak committed:
refactor(service-worker): rename `RegistrationOptions` to `SwRegistrationOptions`

This is in preparation of making `RegistrationOptions` part of the
public API (in a subsequent commit).

gkalpak committed:
docs(service-worker): improve `SwRegistrationOptions` docs and add example

gkalpak committed:
test(service-worker): expand `SwRegistrationOptions` tests and move to separate file

gkalpak committed:
fixup! feat(service-worker): allow configuring when the SW is registered

gkalpak commented:
merge-assistance: Requires manualt CLA verification.
